src/install_pypi/utils.rs

```diff
@@ -0,0 +1,63 @@
+use std::{borrow::Cow, str::FromStr, time::Duration};
+
+use miette::IntoDiagnostic;
+use url::Url;
+use uv_cache::{ArchiveTarget, ArchiveTimestamp};
+use uv_distribution_types::InstalledDist;
+
+pub fn elapsed(duration: Duration) -> String {
+    let secs = duration.as_secs();
+
+    if secs >= 60 {
+        format!("{}m {:02}s", secs / 60, secs % 60)
+    } else if secs > 0 {
+        format!("{}.{:02}s", secs, duration.subsec_nanos() / 10_000_000)
+    } else {
+        format!("{}ms", duration.subsec_millis())
+    }
+}
+
+/// Check if the url is a direct url
+/// Files, git, are direct urls
+/// Direct urls to wheels or sdists are prefixed with a `direct` scheme
+/// by us when resolving the lock file
+pub fn is_direct_url(url_scheme: &str) -> bool {
+    url_scheme == "file"
+        || url_scheme == "git+http"
+        || url_scheme == "git+https"
+        || url_scheme == "git+ssh"
+        || url_scheme.starts_with("direct")
+}
+
+/// Strip of the `direct` scheme from the url if it is there
+pub fn strip_direct_scheme(url: &Url) -> Cow<'_, Url> {
+    url.as_ref()
+        .strip_prefix("direct+")
+        .and_then(|str| Url::from_str(str).ok())
+        .map(Cow::Owned)
+        .unwrap_or(Cow::Borrowed(url))
+}
+
+/// Check freshness of a locked url against an installed dist
+pub fn check_url_freshness(
```

Comment:
do we need them as pub or we can make them pub(crate)?

Reply:
Yeah so I dunno, I recently decided to do it on module level, look at `mod.rs` of that module, then If I want to change the visibility I only need to do it in one place. WDYT?